### PR TITLE
Fix DistinctCount for timestamps with time zone

### LINKED ISSUE #10042: Regression: group by count distinct doesn't work for timestamps with time zone
### Describe the bug

I didn't check in which version it regressed, but it worked in v34 and is broken in v37.
It looks like the regression was caused by the specialization of the `DistinctCount` implementation of `AggregateExpr`. 

### To Reproduce

```sql
create table t_source
as values
 ('2018-11-13T17:11:10.011375885995', 'Row 0', 'X'),
 ('2011-12-13T11:13:10.12345',        'Row 1', 'X'),
 (null,                               'Row 2', 'Y'),
 ('2021-01-01T05:11:10.432',          'Row 3', 'Y');

create table t as
select
  arrow_cast(column1, 'Timestamp(Second, Some("UTC"))') as secs_utc,
  column2 as names,
  column3 as tag
from t_source;

SELECT tag, count(DISTINCT secs_utc) FROM t GROUP BY tag ORDER BY tag;
```

### Expected behavior

It should work. The error we get is:

> External error: query failed: DataFusion error: External error: External error: Arrow error: Invalid argument error: column types must match schema types, expected List(Field { name: "item", data_type: Timestamp(Second, Some("UTC")), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }) but found List(Field { name: "item", data_type: Timestamp(Second, None), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }) at column index 1

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by joroKr21**
## Which issue does this PR close?

Closes #10042

## Rationale for this change

Fixing a regression.

## What changes are included in this PR?

Preserve the original data type in the aggregation state.

## Are these changes tested?

Tested in SLT.

## Are there any user-facing changes?

Internal changes only.

### CODE REVIEW

File: datafusion/physical-expr/src/aggregate/count_distinct/mod.rs [hunk 1]
```diff
@@ -109,12 +109,14 @@ impl AggregateExpr for DistinctCount {
...
-                Box::new(PrimitiveDistinctCountAccumulator::<Decimal256Type>::new())
-            }
+            dt @ Decimal128(_, _) => Box::new(
+                PrimitiveDistinctCountAccumulator::<Decimal128Type>::new()
+                    .with_data_type(dt.clone()),
+            ),
+            dt @ Decimal256(_, _) => Box::new(
+                PrimitiveDistinctCountAccumulator::<Decimal256Type>::new()
+                    .with_data_type(dt.clone()),
+            ),
```
Comment (jayzhan211) [inline datafusion/physical-expr/src/aggregate/count_distinct/mod.rs:119]: Can you add the test for decimal too?

Comment (joroKr21) [reply datafusion/physical-expr/src/aggregate/count_distinct/mod.rs:119]: Sure, added 👍

Comment (jayzhan211) [reply datafusion/physical-expr/src/aggregate/count_distinct/mod.rs:119]: I expect the test that shows the change is needed. If this change is removed, it failed.

Comment (jayzhan211) [reply datafusion/physical-expr/src/aggregate/count_distinct/mod.rs:119]: It is clear for me the change for timestamp with timezone, but I'm not sure about the decimal one.

Comment (joroKr21) [reply datafusion/physical-expr/src/aggregate/count_distinct/mod.rs:119]: I can show that for Decimal128 (not sure why it doesn't trigger for Decimal256). But it's better to be consistent, losing the data type might lead to unexpected consequences.

**Review by alamb — APPROVED** [review]
Thank you @joroKr21  and @jayzhan211

File: datafusion/physical-expr/src/aggregate/count_distinct/mod.rs [hunk 2]
```diff
@@ -130,18 +132,22 @@ impl AggregateExpr for DistinctCount {
...
-            Timestamp(Millisecond, _) => Box::new(PrimitiveDistinctCountAccumulator::<
-                TimestampMillisecondType,
-            >::new()),
-            Timestamp(Nanosecond, _) => Box::new(PrimitiveDistinctCountAccumulator::<
-                TimestampNanosecondType,
-            >::new()),
-            Timestamp(Second, _) => {
-                Box::new(PrimitiveDistinctCountAccumulator::<TimestampSecondType>::new())
-            }
+            dt @ Timestamp(Microsecond, _) => Box::new(
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/count_distinct/mod.rs:135]: I wonder if we should defensively always set `with_data_type` to `PrimitiveDistinctCountAccumulator` 🤔 

I tried to come up with other `DataType` that doesn't have a 1:1 mapping to its `ArrowPrimitiveType` and I don't think there is one, but always supplying the `data_type` might be a more defensive pattern

**Review by Dandandan — APPROVED** [review]
Thank you @joroKr21

File: datafusion/physical-expr/src/aggregate/count_distinct/mod.rs (hunk 2 above)
Comment (joroKr21) [reply datafusion/physical-expr/src/aggregate/count_distinct/mod.rs:135]: I was wondering the same. Let me know what you prefer?

Comment (alamb) [reply datafusion/physical-expr/src/aggregate/count_distinct/mod.rs:135]: How about we do it as a follow on PR? I would like to merge this one asap as it fixes a regression

**Pull request merged**